Src/AutoFixture.xUnit.net/FreezeAttribute.cs

```diff
@@ -0,0 +1,145 @@
+using System;
+using System.Reflection;
+using Ploeh.AutoFixture.Kernel;
+
+namespace Ploeh.AutoFixture.Xunit
+{
+    /// <summary>
+    /// An attribute that can be applied to parameters in an <see cref="AutoDataAttribute"/>-driven
+    /// Theory to indicate that the parameter value should be frozen and used to satisfy any requests
+    /// made through the same <see cref="IFixture"/> and that match a set of criteria.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false)]
+    public sealed class FreezeAttribute : CustomizeAttribute
```

Comment:
Valid points, all.

When I preach SOLID, I always say that the OCP means that you can't go back and change an existing type. If I should practice what I preach, then you've already pointed to the correct solution: leave `[Frozen]` as is, and introduce a new type with the desired behaviour. Possibly, sometime in that process, we could deprecate `[Frozen]`.

My main concern with that approach is that we'd confuse the hell out of users. Even if we put an `[Obsolete]` attribute on `[Frozen]`, with a nice message pointing to the new attribute, I'm afraid that many people would just give up at that point.

The only *consistent* complaint I hear about AutoFixture is that it's too complicated. Your proposed solution, while true to SOLID, doesn't make AutoFixture less complicated.

This is one of the times when it'd be really nice with some metrics about how people use AutoFixture... However, in lack of hard data, let's attempt to look at some incidental points.

The `[Frozen]` attribute has been around since August 2010(!). The `As` property has been around since December 2011, so that's also a long time.

However, [a Google search](https://www.google.com/search?q=AutoFixture%20Frozen%20As) from inside of my Filter Bubble only shows a single hit on the `As` keyword, among the many articles detailing the `[Frozen]` attribute, so I don't think this particular feature is well-known.

The [only hit I've found](http://stackoverflow.com/a/15447124/126014) is one where, as far as I can tell, `Matching.ImplementedInterfaces` would solve the problem nicely.

You are still correct that if we deprecate the `As` property, you *can* come up with examples we can no longer support with the `[Frozen]` attribute. The question is whether anyone use it that way at all.

Personally, I'd be ready to take the bet that no-one would notice if we pulled support for that particular corner case. Even if there's a single developer using it like that somewhere, it's still possible to freeze specimens like that imperatively using `Inject<T>` - you just can't do it declaratively any longer.

Now, if we knew that there are thousands of users using `As` to handle deep inheritance hierarchies, my position would be different, but although I don't have hard data, I don't think that's the case. Hence, my position is still that I'd prefer to deprecate the `As` property, and add the new feature to the existing `[Frozen]` attribute.